File: grooves-core/src/main/java/com/github/rahulsom/grooves/functions/EventHandler.java
package com.github.rahulsom.grooves.functions;

import com.github.rahulsom.grooves.EventApplyOutcome;
import com.github.rahulsom.grooves.logging.Trace;

/**
 * Applies the event to a snapshot.
 */
public interface EventHandler<EventT, SnapshotT> {
    /**
     * Applies the given event to the provided snapshot.
     *
     * @param event the event to apply to the snapshot
     * @param snapshot the snapshot to modify with the event
     * @return the outcome indicating whether to continue processing or return the snapshot
     */
    @Trace
    EventApplyOutcome invoke(
            EventT event,
            SnapshotT snapshot
    );
}

File: grooves-core/src/main/kotlin/com/github/rahulsom/grooves/functions/EventHandler.kt
package com.github.rahulsom.grooves.functions

import com.github.rahulsom.grooves.EventApplyOutcome
import com.github.rahulsom.grooves.logging.Trace

/**
 * Applies the event to a snapshot.
 */
interface EventHandler<Event, Snapshot> {
    @Trace
    fun invoke(event: Event, snapshot: Snapshot): EventApplyOutcome
}